File: day-06/part-2/skasch.go
package main

import (
	"fmt"
	"io/ioutil"
	"log"
	"os"
	"strconv"
	"strings"
	"time"
)

const CYCLE = 7
const NEW = 2
const DAYS = 256

func parse(s string) [CYCLE + NEW]int {
	strs := strings.Split(s, ",")
	mem := [CYCLE + NEW]int{0}
	for _, str := range strs {
		v, err := strconv.Atoi(str)
		if err != nil {
			log.Print(err)
		}
		mem[v]++
	}
	return mem
}

func run(s string) int {
	// Your code goes here
	mem := parse(s)
	news := [NEW]int{0}
	for day := CYCLE; day < DAYS; day += CYCLE {
		for i := 0; i < NEW; i++ {
			news[i] = mem[i+CYCLE]
			mem[i+CYCLE] = 0
		}
		for i := CYCLE - 1; i >= 0; i-- {
			mem[i+2] += mem[i]
		}
		for i := 0; i < NEW; i++ {
			mem[i] += news[i]
		}
	}
	res := 0
	for idx := 0; idx < (DAYS % CYCLE); idx++ {
		res += 2 * mem[idx]
	}
	for idx := (DAYS % CYCLE); idx < CYCLE+NEW; idx++ {
		res += mem[idx]
	}
	return res
}

func main() {
	// Uncomment this line to disable garbage collection
	// debug.SetGCPercent(-1)

	var input []byte
	var err error
	if len(os.Args) > 1 {
		// Read input from file for local debugging
		input, err = ioutil.ReadFile(os.Args[1])
		if err != nil {
			panic(err)
		}
		// Remove extra newline
		input = input[:len(input)-1]
	} else {
		// Read input from stdin
		input, err = ioutil.ReadAll(os.Stdin)
		if err != nil {
			panic(err)
		}
	}

	// Start resolution
	start := time.Now()
	result := run(string(input))

	// Print result
	fmt.Printf("_duration:%f\n", time.Since(start).Seconds()*1000)
	fmt.Println(result)
}

File: day-06/part-2/skasch.py

import collections
from typing import List
from tool.runners.python import SubmissionPy


DAYS = 256
CYCLE = 7
NEW = 2


def parse(s: str) -> List[str]:
    return s.split(",")


class SkaschSubmission(SubmissionPy):
    def run(self, s):
        """
        :param s: input in string format
        :return: solution flag
        """
        # Your code goes here
        cnt = collections.Counter([int(v) for v in parse(s)])
        d = DAYS
        while d >= CYCLE:
            tmp_cnt = collections.Counter()
            for v in range(CYCLE, CYCLE + NEW):
                tmp_cnt[v - CYCLE] += cnt[v]
                del cnt[v]
            cnt += collections.Counter({v + 2: n for v, n in cnt.items()})
            cnt += tmp_cnt
            d -= CYCLE
        return sum(n * (1 if v >= d else 2) for v, n in cnt.items())


def test_skasch():
    """
    Run `python -m pytest ./day-06/part-1/skasch.py` to test the submission.
    """
    assert (
        SkaschSubmission().run(
            """
3,4,3,1,2
""".strip()
        )
        == 26984457539
    )
